File: Source/AlphaTab/Platform/Svg/SvgCanvas.cs
using AlphaTab.Collections;
using AlphaTab.Platform.Model;
using AlphaTab.Rendering.Glyphs;

namespace AlphaTab.Platform.Svg
{
    /// <summary>
    ///  A canvas implementation storing SVG data
    /// </summary>
    internal abstract class SvgCanvas : ICanvas
    {
        protected const float BlurCorrection = 0;

        protected StringBuilder Buffer;
        private StringBuilder _currentPath;
        private bool _currentPathIsEmpty;

        public Color Color { get; set; }
        public float LineWidth { get; set; }
        public Font Font { get; set; }
        public TextAlign TextAlign { get; set; }
        public TextBaseline TextBaseline { get; set; }
        public Settings Settings { get; set; }

        public SvgCanvas()
        {
            _currentPath = new StringBuilder();
            _currentPathIsEmpty = true;
            Color = new Color(255, 255, 255);
            LineWidth = 1;
            Font = new Font("Arial", 10);
            TextAlign = TextAlign.Left;
            TextBaseline = TextBaseline.Top;
        }

        public virtual void BeginRender(float width, float height)
        {
            Buffer = new StringBuilder();

            Buffer.Append("<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" width=\"");
            Buffer.Append(width);
            Buffer.Append("px\" height=\"");
            Buffer.Append((int)height);
            Buffer.Append("px\" class=\"alphaTabSurfaceSvg\">\n");
            _currentPath = new StringBuilder();
            _currentPathIsEmpty = true;
        }

        public void BeginGroup(string identifier)
        {
            Buffer.Append("<g class=\"" + identifier + "\">");
        }

        public void EndGroup()
        {
            Buffer.Append("</g>");
        }

        public virtual object EndRender()
        {
            Buffer.Append("</svg>");
            return Buffer.ToString();
        }

        public void FillRect(float x, float y, float w, float h)
        {
            if (w > 0)
            {
                Buffer.Append("<rect x=\"" + ((int)x - BlurCorrection) + "\" y=\"" + ((int)y - BlurCorrection) +
                              "\" width=\"" + w + "\" height=\"" + h + "\" fill=\"" + Color.Rgba + "\" />\n");
            }
        }

        public void StrokeRect(float x, float y, float w, float h)
        {
            Buffer.Append("<rect x=\""
                          + ((int)x - BlurCorrection)
                          + "\" y=\""
                          + ((int)y - BlurCorrection)
                          + "\" width=\""
                          + w
                          + "\" height=\""
                          + h
                          + "\" stroke=\"" + Color.Rgba + "\"");
            if (LineWidth != 1)
            {
                Buffer.Append(" stroke-width=\"" + LineWidth + "\"");
            }

            Buffer.Append(" fill=\"transparent\" />\n");
        }

        public void BeginPath()
        {
        }

        public void ClosePath()
        {
            _currentPath.Append(" z");
        }

        public void MoveTo(float x, float y)
        {
            _currentPath.Append(" M" + (x - BlurCorrection) + "," + (y - BlurCorrection));
        }

        public void LineTo(float x, float y)
        {
            _currentPathIsEmpty = false;
            _currentPath.Append(" L" + (x - BlurCorrection) + "," + (y - BlurCorrection));
        }

        public void QuadraticCurveTo(float cpx, float cpy, float x, float y)
        {
            _currentPathIsEmpty = false;
            _currentPath.Append(" Q" + cpx + "," + cpy + "," + x + "," + y);
        }

        public void BezierCurveTo(float cp1X, float cp1Y, float cp2X, float cp2Y, float x, float y)
        {
            _currentPathIsEmpty = false;
            _currentPath.Append(" C" + cp1X + "," + cp1Y + "," + cp2X + "," + cp2Y + "," + x + "," + y);
        }

        public void FillCircle(float x, float y, float radius)
        {
            _currentPathIsEmpty = false;
            //
            // M0,250 A1,1 0 0,0 500,250 A1,1 0 0,0 0,250 z
            _currentPath.Append(" M" + (x - radius) + "," + y + " A1,1 0 0,0 " + (x + radius) + "," + y +
                                " A1,1 0 0,0 " + (x - radius) + "," + y + " z");
            Fill();
        }

        public void Fill()
        {
            if (!_currentPathIsEmpty)
            {
                Buffer.Append("<path d=\"" + _currentPath + "\"");
                if (Color.Rgba != Color.BlackRgb)
                {
                    Buffer.Append(" fill=\"" + Color.Rgba + "\"");
                }

                Buffer.Append(" style=\"stroke: none\"/>");
            }

            _currentPath = new StringBuilder();
            _currentPathIsEmpty = true;
        }

        public void Stroke()
        {
            if (!_currentPathIsEmpty)
            {
                var s = "<path d=\"" + _currentPath + "\" stroke=\"" + Color.Rgba + "\"";
                if (LineWidth != 1)
                {
                    s += " stroke-width=\"" + LineWidth + "\"";
                }

                s += " style=\"fill: none\" />";
                Buffer.Append(s);
            }

            _currentPath = new StringBuilder();
            _currentPathIsEmpty = true;
        }

        public void FillText(string text, float x, float y)
        {
            if (text == "")
            {
                return;
            }

            var s = "<text x=\"" + (int)x + "\" y=\"" + (int)y + "\" style=\"stroke: none; font:" +
                    Font.ToCssString(Settings.Display.Scale) + "\" "
                    + " dominant-baseline=\"" + GetSvgBaseLine() + "\"";
            if (Color.Rgba != Color.BlackRgb)
            {
                s += " fill=\"" + Color.Rgba + "\"";
            }

            if (TextAlign != TextAlign.Left)
            {
                s += " text-anchor=\"" + GetSvgTextAlignment(TextAlign) + "\"";
            }

            s += ">" + text + "</text>";
            Buffer.Append(s);
        }

        protected string GetSvgTextAlignment(TextAlign textAlign)
        {
            switch (textAlign)
            {
                case TextAlign.Left: return "start";
                case TextAlign.Center: return "middle";
                case TextAlign.Right: return "end";
            }

            return "";
        }

        private string GetSvgBaseLine()
        {
            switch (TextBaseline)
            {
                case TextBaseline.Top: return "hanging";
                case TextBaseline.Middle: return "middle";
                case TextBaseline.Bottom: return "bottom";
                default: return "";
            }
        }

        public float MeasureText(string text)
        {
            if (string.IsNullOrEmpty(text))
            {
                return 0;
            }

            return FontSizes.MeasureString(text, Font.Family, Font.Size, Font.Style);
        }

        public abstract void FillMusicFontSymbol(
            float x,
            float y,
            float scale,
            MusicFontSymbol symbol,
            bool centerAtPosition = false);

        public abstract void FillMusicFontSymbols(
            float x,
            float y,
            float scale,
            MusicFontSymbol[] symbols,
            bool centerAtPosition = false);

        public virtual object OnRenderFinished()
        {
            // nothing to do
            return null;
        }

        public void BeginRotate(float centerX, float centerY, float angle)
        {
            Buffer.Append("<g transform=\"translate(" + centerX + " ," + centerY + ") rotate( " + angle + ")\">");
        }

        public void EndRotate()
        {
            Buffer.Append("</g>");
        }
    }
}

File: src/platform/svg/SvgCanvas.ts
import { Color } from '@src/model/Color';
import { Font, FontStyle } from '@src/model/Font';
import { type ICanvas, TextAlign, TextBaseline, MeasuredText } from '@src/platform/ICanvas';
import { FontSizes } from '@src/platform/svg/FontSizes';
import type { MusicFontSymbol } from '@src/model/MusicFontSymbol';
import type { Settings } from '@src/Settings';

/**
 * A canvas implementation storing SVG data
 */
export abstract class SvgCanvas implements ICanvas {
    protected buffer: string = '';

    private _currentPath: string = '';
    private _currentPathIsEmpty: boolean = true;

    public scale = 1;

    public color: Color = new Color(255, 255, 255, 0xff);
    public lineWidth: number = 1;
    public font: Font = new Font('Arial', 10, FontStyle.Plain);
    public textAlign: TextAlign = TextAlign.Left;
    public textBaseline: TextBaseline = TextBaseline.Top;
    public settings!: Settings;

    public destroy() {}

    public beginRender(width: number, height: number): void {
        this.scale = this.settings.display.scale;
        this.buffer = `<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="${width | 0}px" height="${
            height | 0
        }px" class="at-surface-svg">\n`;
        this._currentPath = '';
        this._currentPathIsEmpty = true;
        this.textBaseline = TextBaseline.Top;
    }

    public beginGroup(identifier: string): void {
        this.buffer += `<g class="${identifier}">`;
    }

    public endGroup(): void {
        this.buffer += '</g>';
    }

    public endRender(): unknown {
        this.buffer += '</svg>';
        return this.buffer;
    }

    public fillRect(x: number, y: number, w: number, h: number): void {
        if (w > 0) {
            this.buffer += `<rect x="${(x * this.scale) | 0}" y="${(y * this.scale) | 0}" width="${
                w * this.scale
            }" height="${h * this.scale}" fill="${this.color.rgba}" />\n`;
        }
    }

    public strokeRect(x: number, y: number, w: number, h: number): void {
        const blurOffset = (this.lineWidth * this.scale) % 2 === 0 ? 0 : 0.5;
        this.buffer += `<rect x="${((x * this.scale) | 0) + blurOffset}" y="${((y * this.scale) | 0) + blurOffset}" width="${
            w * this.scale
        }" height="${h * this.scale}" stroke="${this.color.rgba}"`;
        if (this.lineWidth !== 1) {
            this.buffer += ` stroke-width="${this.lineWidth * this.scale}"`;
        }
        this.buffer += ' fill="transparent" />\n';
    }

    public beginPath(): void {
        // nothing to do
    }

    public closePath(): void {
        this._currentPath += ' z';
    }

    public moveTo(x: number, y: number): void {
        this._currentPath += ` M${x * this.scale},${y * this.scale}`;
    }

    public lineTo(x: number, y: number): void {
        this._currentPathIsEmpty = false;
        this._currentPath += ` L${x * this.scale},${y * this.scale}`;
    }

    public quadraticCurveTo(cpx: number, cpy: number, x: number, y: number): void {
        this._currentPathIsEmpty = false;
        this._currentPath += ` Q${cpx * this.scale},${cpy * this.scale},${x * this.scale},${y * this.scale}`;
    }

    public bezierCurveTo(cp1X: number, cp1Y: number, cp2X: number, cp2Y: number, x: number, y: number): void {
        this._currentPathIsEmpty = false;
        this._currentPath += ` C${cp1X * this.scale},${cp1Y * this.scale},${cp2X * this.scale},${cp2Y * this.scale},${
            x * this.scale
        },${y * this.scale}`;
    }

    public fillCircle(x: number, y: number, radius: number): void {
        this._currentPathIsEmpty = false;

        x *= this.scale;
        y *= this.scale;
        radius *= this.scale;

        //
        // M0,250 A1,1 0 0,0 500,250 A1,1 0 0,0 0,250 z
        this._currentPath += ` M${x - radius},${y} A1,1 0 0,0 ${x + radius},${y} A1,1 0 0,0 ${x - radius},${y} z`;
        this.fill();
    }

    public strokeCircle(x: number, y: number, radius: number): void {
        this._currentPathIsEmpty = false;

        x *= this.scale;
        y *= this.scale;
        radius *= this.scale;

        //
        // M0,250 A1,1 0 0,0 500,250 A1,1 0 0,0 0,250 z
        this._currentPath += ` M${x - radius},${y} A1,1 0 0,0 ${x + radius},${y} A1,1 0 0,0 ${x - radius},${y} z`;
        this.stroke();
    }

    public fill(): void {
        if (!this._currentPathIsEmpty) {
            this.buffer += `<path d="${this._currentPath}"`;
            if (this.color.rgba !== '#000000') {
                this.buffer += ` fill="${this.color.rgba}"`;
            }
            this.buffer += ' style="stroke: none"/>';
        }
        this._currentPath = '';
        this._currentPathIsEmpty = true;
    }

    public stroke(): void {
        if (!this._currentPathIsEmpty) {
            let s: string = `<path d="${this._currentPath}" stroke="${this.color.rgba}"`;
            if (this.lineWidth !== 1 || this.scale !== 1) {
                s += ` stroke-width="${this.lineWidth * this.scale}"`;
            }
            s += ' style="fill: none" />';
            this.buffer += s;
        }
        this._currentPath = '';
        this._currentPathIsEmpty = true;
    }

    public fillText(text: string, x: number, y: number): void {
        if (text === '') {
            return;
        }
        let s: string = `<text x="${(x * this.scale) | 0}" y="${
            (y * this.scale) | 0
        }" style='stroke: none; font:${this.font.toCssString(this.settings.display.scale)}; ${this.getSvgBaseLine()}'`;
        if (this.color.rgba !== '#000000') {
            s += ` fill="${this.color.rgba}"`;
        }
        if (this.textAlign !== TextAlign.Left) {
            s += ` text-anchor="${this.getSvgTextAlignment(this.textAlign)}"`;
        }
        s += `>${SvgCanvas.escapeText(text)}</text>`;
        this.buffer += s;
    }

    private static escapeText(text: string) {
        return text
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#39;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
    }

    protected getSvgTextAlignment(textAlign: TextAlign): string {
        switch (textAlign) {
            case TextAlign.Left:
                return 'start';
            case TextAlign.Center:
                return 'middle';
            case TextAlign.Right:
                return 'end';
        }
        return '';
    }

    protected getSvgBaseLine(): string {
        switch (this.textBaseline) {
            case TextBaseline.Top:
                return 'dominant-baseline: hanging';
            case TextBaseline.Bottom:
                return 'dominant-baseline: ideographic';
            // case TextBaseline.Middle:
            default:
                // middle is set as default on the SVG tag via css
                return '';
        }
    }

    public measureText(text: string) {
        if (!text) {
            return new MeasuredText(0, 0);
        }
        return FontSizes.measureString(
            text,
            this.font.families,
            this.font.size * this.scale,
            this.font.style,
            this.font.weight
        );
    }

    public abstract fillMusicFontSymbol(
        x: number,
        y: number,
        relativeScale: number,
        symbol: MusicFontSymbol,
        centerAtPosition?: boolean
    ): void;

    public abstract fillMusicFontSymbols(
        x: number,
        y: number,
        relativeScale: number,
        symbols: MusicFontSymbol[],
        centerAtPosition?: boolean
    ): void;

    public onRenderFinished(): unknown {
        // nothing to do
        return null;
    }

    public beginRotate(centerX: number, centerY: number, angle: number): void {
        this.buffer += `<g transform="translate(${centerX * this.scale} ,${centerY * this.scale}) rotate( ${angle})">`;
    }

    public endRotate(): void {
        this.buffer += '</g>';
    }
}
